op-program/host/prefetcher/reexec.go

```diff
@@ -19,19 +20,27 @@ type ProgramExecutor interface {
 // re-derive the block.
 func (p *Prefetcher) nativeReExecuteBlock(
 	ctx context.Context, agreedBlockHash, blockHash common.Hash, chainID uint64) error {
-	// Avoid retries as the block may not be canonical and unavailable
-
+	// Avoid using the retrying source to prevent indefinite retries as the block may not be canonical and unavailable
 	source, err := p.l2Sources.ForChainIDWithoutRetries(chainID)
 	if err != nil {
 		return err
 	}
-	_, _, err = source.InfoAndTxsByHash(ctx, blockHash)
-	if err == nil {
+	notFound, err := retry.Do(ctx, maxAttempts, retry.Exponential(), func() (bool, error) {
+		_, _, err := source.InfoAndTxsByHash(ctx, blockHash)
+		if errors.Is(err, ethereum.NotFound) {
+			return true, nil
+		}
+		if err != nil {
+			p.logger.Warn("Failed to retrieve l2 info and txs", "hash", blockHash, "err", err)
+		}
+		return false, err
+	})
+	if !notFound && err == nil {
 		// we already have the data needed for the program to re-execute
 		return nil
 	}
-	if !errors.Is(err, ethereum.NotFound) {
-		p.logger.Error("Failed to fetch block", "block_hash", blockHash, "err", err)
+	if notFound {
```

Comment:
What do we do when we get an error that isn't not found? ie `!notFound && err != nil`?

Reply:
We treat that as if the block were unavailable and rebuild it. I'll add a comment.

Reply:
The `maxAttempt` retry policy ensures that never happens as it retries indefinitely. The `!notFound && err != nil` case case is handled as a fallback just in case the retry policy changes.